### PR TITLE
Server per test

### PR DESCRIPTION
**Pull request by BrennanConroy**
Lets hash out the design.

Currently is working and an example of test log is below:

```
[0.001s] [TestLifetime] [Information] Starting test CanSendAndReceiveMessage_json_ServerSentEvents_default at 2018-11-02T16:17:51
[0.002s] [Microsoft.AspNetCore.SignalR.Tests.ServerFixture] [Information] Starting test server...
[0.004s] [Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager] [Information] User profile is available. Using 'C:\Users\brecon\AppData\Local\ASP.NET\DataProtection-Keys' as key repository and Windows DPAPI to encrypt keys at rest.
[0.027s] [Microsoft.AspNetCore.SignalR.Tests.ServerFixture] [Information] Test Server started
[0.031s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request starting HTTP/1.1 POST http://127.0.0.1:39919/default/negotiate  0
[0.031s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request finished in 0.6421ms 200 application/json
[0.032s] [Microsoft.AspNetCore.Http.Connections.Client.Internal.ServerSentEventsTransport] [Information] e8GsoqDPQ1_44SDhQHsEMw - Starting transport. Transfer mode: Text.
[0.033s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request starting HTTP/1.1 GET http://127.0.0.1:39919/default?id=e8GsoqDPQ1_44SDhQHsEMw  
[0.033s] [Microsoft.AspNetCore.Http.Connections.Client.HttpConnection] [Information] e8GsoqDPQ1_44SDhQHsEMw - HttpConnection Started.
[0.034s] [Microsoft.AspNetCore.SignalR.Client.HubConnection] [Information] e8GsoqDPQ1_44SDhQHsEMw - Using HubProtocol 'json v1'.
[0.035s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request starting HTTP/1.1 POST http://127.0.0.1:39919/default?id=e8GsoqDPQ1_44SDhQHsEMw  32
[0.035s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request finished in 0.2683ms 200 text/plain
[0.035s] [Microsoft.AspNetCore.SignalR.HubConnectionContext] [Information] e8GsoqDPQ1_44SDhQHsEMw - e8GsoqDPQ1_44SDhQHsEMw - Completed connection handshake. Using HubProtocol 'json'.
[0.035s] [Microsoft.AspNetCore.SignalR.Client.HubConnection] [Information] e8GsoqDPQ1_44SDhQHsEMw - HubConnection started.
[0.036s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request starting HTTP/1.1 POST http://127.0.0.1:39919/default?id=e8GsoqDPQ1_44SDhQHsEMw  11
[0.036s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request finished in 0.1579ms 200 text/plain
[0.036s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request starting HTTP/1.1 POST http://127.0.0.1:39919/default?id=e8GsoqDPQ1_44SDhQHsEMw  70
[0.036s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request finished in 0.1762ms 200 text/plain
[0.037s] [Microsoft.AspNetCore.Http.Connections.Client.Internal.ServerSentEventsTransport] [Information] e8GsoqDPQ1_44SDhQHsEMw - Transport is stopping.
[0.037s] [Microsoft.AspNetCore.Http.Connections.Client.HttpConnection] [Information] e8GsoqDPQ1_44SDhQHsEMw - HttpConnection Disposed.
[0.037s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request finished in 4.5803ms 200 text/event-stream
[0.038s] [Microsoft.AspNetCore.SignalR.Tests.ServerFixture] [Information] Shutting down test server
[0.038s] [Microsoft.AspNetCore.SignalR.Tests.ServerFixture] [Information] Test server shut down
[0.038s] [TestLifetime] [Information] Finished test CanSendAndReceiveMessage_json_ServerSentEvents_default in 0.0365368s

```

### CODE REVIEW

File: test/Microsoft.AspNetCore.SignalR.Tests.Utils/VerifiableServerLoggedTest.cs [hunk 1]
```diff
@@ -44,6 +43,20 @@ public VerifiableServerLoggedTest(ServerFixture serverFixture, ITestOutputHelper
             };
         }
 
+        public IDisposable StartVerifiableServerLog<T>(out ILoggerFactory loggerFactory, out ServerFixture<T> serverFixture, LogLevel minLogLevel, [CallerMemberName] string testName = null, Func<WriteContext, bool> expectedErrorsFilter = null) where T : class
```
Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.Tests.Utils/VerifiableServerLoggedTest.cs]:
```suggestion
        public IDisposable StartServer<T>(out ILoggerFactory loggerFactory, out ServerFixture<T> serverFixture, LogLevel minLogLevel, [CallerMemberName] string testName = null, Func<WriteContext, bool> expectedErrorsFilter = null) where T : class
```

File: test/Microsoft.AspNetCore.SignalR.Tests.Utils/VerifiableServerLoggedTest.cs [hunk 2]
```diff
@@ -44,6 +43,20 @@ public VerifiableServerLoggedTest(ServerFixture serverFixture, ITestOutputHelper
...
+        {
+            var disposable = base.StartVerifiableLog(out loggerFactory, minLogLevel, testName, ResolveExpectedErrorsFilter(expectedErrorsFilter));
+            serverFixture = new ServerFixture<T>(loggerFactory);
+            return new MultiDisposable(serverFixture, disposable);
+        }
+
+        public IDisposable StartVerifiableServerLog<T>(out ILoggerFactory loggerFactory, out ServerFixture<T> serverFixture, [CallerMemberName] string testName = null, Func<WriteContext, bool> expectedErrorsFilter = null) where T : class
```
Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.Tests.Utils/VerifiableServerLoggedTest.cs]:
```suggestion
        public IDisposable StartServer<T>(out ILoggerFactory loggerFactory, out ServerFixture<T> serverFixture, [CallerMemberName] string testName = null, Func<WriteContext, bool> expectedErrorsFilter = null) where T : class
```

**Comment by anurse** [comment]
Looking snazzy AF. Can you get some rough test run-time comparison numbers from before and after (with a little bit of warm-up)?

**Comment by BrennanConroy** [comment]
Microsoft.AspNetCore.SignalR.Client.FunctionalTests
--
Before: Core 59 seconds, Full Framework 1 minute
After: Core 31.5 seconds, Full Framework 36.9 seconds

**Comment by anurse** [comment]
Seems super weird that it's *faster* after this change... I'd expect it to be slightly slower or about the same...

**Comment by BrennanConroy** [comment]
It has to be parallelization

**Comment by BrennanConroy** [comment]
@muratg FYI, test only changes for release/2.2

**Comment by anurse** [comment]
> It has to be parallelization

Of course! Missed that part ;)

**Pull request merged**